### ISSUE TITLE
Listings Management Field Set: Ranking and Results

### ISSUE THREAD

**Issue by slowbot** [bloom-housing/bloom#1178]
**Goal**: Leasing agent should be able to specify details about lotteries and waitlists

**Fields**:
Field Name | Entity Model | Type | Current State | Required
------------ | ------------- | --- | ----- | -----
Listing Type |  | string null | No | Yes
Lottery Date | programRules | string null | Yes | If enabled
Is Waitlist Open | isWaitlistOpen | boolean | Yes | No
Display Waitlist | displayWaitlist | string null | Updates needed | Yes
Max Waitlist Size | displayWaitlist | string null | Updates needed | If enabled
Current Waitlist Size | displayWaitlist | string null | Updates needed| If enabled
Open Spots on Waitlist | displayWaitlist | string null | Updates needed| If enabled
Lottery Results Date | | | ? | No
Lottery Results | | | ? | No

**Engineering Notes**
- Listing Type values are First Come First Serve and Lottery
- Display Waitlist should show or hide waitlistMaxSize, waitlistCurrentSize, Open Spots on Waitlist
- 

**Questions**

**Figma**:
Default:
https://www.figma.com/file/Y8bFd0aSz7eYJ2TCNid6s3/bloom-partners?node-id=2721%3A127471

Filled:
https://www.figma.com/file/Y8bFd0aSz7eYJ2TCNid6s3/bloom-partners?node-id=2721%3A127473

**Comment by slowbot** [comment]
@pbn4 Want to review this issue with you and make a plan for new fields in the backend.

**Comment by slowbot** [comment]
@pbn4 

- My mistake on programRules, that was copy and pasted
- I think we are going to provide the option for agents to upload Results as well as link them

**Comment by seanmalbert** [comment]
The only thing about this that's in question is what's going on around results. I think it makes sense to continue to have that as a derived field. What I'm not sure about is if it makes sense to have it derived from events. The lottery is an event, but results can be posted independently of that. So to me, it makes sense that we update it to be derived from "results" itself. Results can either be an uploaded document or a link, or perhaps both? So it probably makes sense to have two separate columns:
1. result_doc_id: should relate to the asset_id (from assets) of the document uploaded to cloudinary, so this is dependent on #1170
2. result_link: string

**Comment by seanmalbert** [comment]
So with the above comment in mind, the proposed changes from my perspective are:
- add a result_doc_id column to listings, this id is referenced from the id of the asset in the assets table (when #1170 is complete)
- add a result_link column to listings for results link
- results and lottery dates can be handled by events, so we would need to update the enum for event types
- add isWaitlistOpen: boolean column - since we're no longer going to have calculated fields around the waitlist, it makes it easier to have this separate, especially if an agent doesn't want to list numbers and they just want to let people know it's open
- add waitlistOpenSpots: number column to listings

@pbn4 , does this make sense to you?

**Comment by seanmalbert** [comment]
I also want to add that we should use both the results and result date event to update the listings status that there are results. And I am open to having result date be its own column, but I think we can keep it as an event with the type enum updated.

**Comment by pbn4** [comment]
> @pbn4 , does this make sense to you?

Yes makes sense, no comments to be added.

> but I think we can keep it as an event with the type enum updated.

Could you elaborate, because I'm unsure what that would look like.

**Comment by pbn4** [comment]
@seanmalbert

**Comment by pbn4** [comment]
@seanmalbert Could you explain why do we need result_link if assets contain file_id, so with a result_doc_id we can easily get the file_id too?

**Comment by seanmalbert** [comment]
@pbn4 ,

> > @pbn4 , does this make sense to you?
> 
> Yes makes sense, no comments to be added.
> 
> > but I think we can keep it as an event with the type enum updated.
> 
> Could you elaborate, because I'm unsure what that would look like.

Right now event types are `openHouse` and `publicLottery`, so we would add to that something like `postResults`, `resultDate` or `results`.

> @seanmalbert Could you explain why do we need result_link if assets contain file_id, so with a result_doc_id we can easily get the file_id too?

My understanding is that we want to accommodate both files and links. So, yes, we could use file_id to store the link, and if we do that, it'd be nice to have a flag so we don't have to check if file_id is a link or a file_id.

> > results and lottery dates can be handled by events, so we would need to update the enum for event types
> 
> I'd rather flatten it judging but what a convoluted frontend logic it has created to determine whether a listing has a lottery/results event.

If there is convoluted logic on the front end, then we can still solve that in the backend with a getter or getters and still keep listing_events, but I'm also cool with flattening it, if it makes querying more efficient.

**Comment by pbn4** [comment]
@seanmalbert Now when I think about it it's not even about querying, but the fact that all of those "events" do not appear to me as a logical list, neither they are displayed as such in the frontend. Input forms for this data is also not a table like form where you have a list view and an add/edit button. I can imagine a scenario where one of them requires additional properties like e.g. open house event manager phone number. I'd go with separating all of those events into separate columns on the Listing model like:

```
open_house jsonb
public_lottery jsonb
results jsonb 
```

that would also make logic determining whether something is defined trivial. 

Let me know what you think. I'm not sure about those being jsonb, but I see no better option at the moment.

**Comment by seanmalbert** [comment]
@pbn4 ,

Coming back around to your last comment, I go back and forth on this. My initial thoughts were aligned with yours, where they don't appear to be a logical list and they're not inputted as such and we should split them out. However, their purpose is similar where they are all `events` of the listing and are used to notify the public of major milestones of the listing. They are all also displayed in a similar way on the public site. All at the top of the right column, at least according to the figma mocks, so they could use the same component, even if they don't now.

I also wanted to move away from the current events jsonb column that's on the listing and use the listing_events table. Of course this doesn't accommodate scenarios where additional information needs to be tied to events, but I'd rather wait to see what real scenarios may come in and perhaps we could make that generic too.

So in short, while input/write may require more logic, the read/display can be simple if kept as these "events". It's possible that we could work with @slowbot to get them inputted in a list/table together, but I think it makes sense as it is with the booleans around lottery and results.

**Comment by pbn4** [comment]
> results and lottery dates can be handled by events, so we would need to update the enum for event types

@seanmalbert Current definition of ListingEventType:

```
export enum ListingEventType {
  openHouse = "openHouse",
  publicLottery = "publicLottery",
  lotteryResults = "lotteryResults",
}
```

do we need to add anything to that?

**Comment by pbn4** [comment]
> Right now event types are openHouse and publicLottery, so we would add to that something like postResults, resultDate or results.

We already have `lotteryResults` in ListingEventType:

```
export enum ListingEventType {
  openHouse = "openHouse",
  publicLottery = "publicLottery",
  lotteryResults = "lotteryResults",
}
```

so I think we do not need to add anything.

**Comment by pbn4** [comment]
> I also want to add that we should use both the results and result date event to update the listings status that there are results. And I am open to having result date be its own column, but I think we can keep it as an event with the type enum updated.

Do you mean that we should use createdAt defined in ListingEvent? (this is currently jsonb so requires making it a separate table now, which is planned but not yet implemented)

**Comment by kathyccheng** [comment]
Closed in #1430
